File: bindings/java-ffm/src/main/java/org/maplibre/nativeffi/internal/callback/ResourceTransformState.java
package org.maplibre.nativeffi.internal.callback;

import java.lang.foreign.Arena;
import java.lang.foreign.MemorySegment;
import java.lang.foreign.ValueLayout;
import java.nio.charset.StandardCharsets;
import java.util.Objects;
import org.maplibre.nativeffi.internal.c.MapLibreNativeC;
import org.maplibre.nativeffi.internal.c.mln_resource_transform;
import org.maplibre.nativeffi.internal.c.mln_resource_transform_callback;
import org.maplibre.nativeffi.internal.c.mln_resource_transform_response;
import org.maplibre.nativeffi.internal.convert.NativeValues;
import org.maplibre.nativeffi.internal.memory.MemoryUtil;
import org.maplibre.nativeffi.resource.ResourceTransformCallback;
import org.maplibre.nativeffi.resource.ResourceTransformRequest;

/** Owns runtime-scoped resource transform callback state. */
public final class ResourceTransformState implements AutoCloseable {
  private final Arena arena;
  private final CallbackLifecycle lifecycle = new CallbackLifecycle();
  private final ResourceTransformCallback callback;
  private final MemorySegment stub;
  private final MemorySegment descriptor;

  public ResourceTransformState(ResourceTransformCallback callback) {
    this.arena = Arena.ofShared();
    this.callback = callback;
    this.stub = mln_resource_transform_callback.allocate(this::invoke, arena);
    this.descriptor = mln_resource_transform.allocate(arena);
    mln_resource_transform.size(descriptor, (int) mln_resource_transform.sizeof());
    mln_resource_transform.callback(descriptor, stub);
    mln_resource_transform.user_data(descriptor, MemorySegment.NULL);
  }

  public MemorySegment descriptor() {
    return descriptor;
  }

  public boolean isCurrentThreadInCallback() {
    return lifecycle.isCurrentThreadInCallback();
  }

  private int invoke(
      MemorySegment userData, int rawKind, MemorySegment url, MemorySegment outResponse) {
    var lease = lifecycle.enter();
    if (lease.isEmpty()) {
      return MapLibreNativeC.MLN_STATUS_NATIVE_ERROR();
    }
    try (var ignored = lease.get()) {
      mln_resource_transform_response.size(
          outResponse, (int) mln_resource_transform_response.sizeof());
      mln_resource_transform_response.url(outResponse, MemorySegment.NULL);
      var replacement =
          callback.transform(
              new ResourceTransformRequest(
                  NativeValues.resourceKind(rawKind), rawKind, MemoryUtil.copyCString(url)));
      Objects.requireNonNull(replacement, "replacement");
      if (replacement.isPresent() && !replacement.get().isEmpty()) {
        return setResponseUrl(outResponse, replacement.get());
      }
      return MapLibreNativeC.MLN_STATUS_OK();
    } catch (IllegalArgumentException error) {
      return MapLibreNativeC.MLN_STATUS_INVALID_ARGUMENT();
    } catch (Throwable ignored) {
      return MapLibreNativeC.MLN_STATUS_NATIVE_ERROR();
    }
  }

  @Override
  public void close() {
    lifecycle.close("Resource transform", arena::close);
  }

  private static int setResponseUrl(MemorySegment outResponse, String value) {
    var bytes = value.getBytes(StandardCharsets.UTF_8);
    try (var responseArena = Arena.ofConfined()) {
      return MapLibreNativeC.mln_resource_transform_response_set_url(
          outResponse, responseArena.allocateFrom(ValueLayout.JAVA_BYTE, bytes), bytes.length);
    }
  }
}

File: bindings/kotlin/src/jvmMain/kotlin/org/maplibre/nativeffi/internal/callback/ResourceTransformState.kt
package org.maplibre.nativeffi.internal.callback

import java.lang.foreign.Arena
import java.lang.foreign.Linker
import java.lang.foreign.MemorySegment
import java.lang.foreign.ValueLayout
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType
import org.maplibre.nativeffi.error.MaplibreStatus
import org.maplibre.nativeffi.internal.c.mln_resource_transform
import org.maplibre.nativeffi.internal.c.mln_resource_transform_callback
import org.maplibre.nativeffi.internal.c.mln_resource_transform_response
import org.maplibre.nativeffi.internal.c.mln_runtime_callback_release
import org.maplibre.nativeffi.internal.lifecycle.HandleLeakCleaner
import org.maplibre.nativeffi.internal.loader.NativeAccess
import org.maplibre.nativeffi.resource.ResourceKind
import org.maplibre.nativeffi.resource.ResourceTransformCallback
import org.maplibre.nativeffi.resource.ResourceTransformRequest

/** Owns runtime-scoped JVM FFM resource transform callback state. */
internal class ResourceTransformState(private val callback: ResourceTransformCallback) :
  AutoCloseable {
  private val arena = Arena.ofAuto()
  private val gate = CallbackGate("resource transform callbacks") {}
  private val stub: MemorySegment
  private val descriptor: MemorySegment

  init {
    val method =
      MethodHandles.lookup()
        .findVirtual(
          ResourceTransformState::class.java,
          "invoke",
          MethodType.methodType(
            Int::class.javaPrimitiveType,
            MemorySegment::class.java,
            Int::class.javaPrimitiveType,
            MemorySegment::class.java,
            MemorySegment::class.java,
          ),
        )
        .bindTo(this)
    stub = Linker.nativeLinker().upcallStub(method, callbackDescriptor, arena)
    descriptor = mln_resource_transform.allocate(arena)
    mln_resource_transform.size(descriptor, mln_resource_transform.sizeof().toInt())
    mln_resource_transform.callback(descriptor, stub)
    mln_resource_transform.user_data(descriptor, MemorySegment.NULL)
    val releaseMethod =
      MethodHandles.lookup()
        .findVirtual(
          ResourceTransformState::class.java,
          "release",
          MethodType.methodType(Void.TYPE, MemorySegment::class.java),
        )
        .bindTo(this)
    mln_resource_transform.release_user_data(
      descriptor,
      Linker.nativeLinker()
        .upcallStub(releaseMethod, mln_runtime_callback_release.descriptor(), arena),
    )
  }

  fun descriptor(): MemorySegment = descriptor

  fun invoke(
    userData: MemorySegment,
    rawKind: Int,
    url: MemorySegment,
    outResponse: MemorySegment,
  ): Int {
    if (outResponse == MemorySegment.NULL) return MaplibreStatus.INVALID_ARGUMENT.nativeCode
    val lease = gate.enter() ?: return MaplibreStatus.INVALID_ARGUMENT.nativeCode
    return try {
      mln_resource_transform_response.size(
        outResponse,
        mln_resource_transform_response.sizeof().toInt(),
      )
      mln_resource_transform_response.url(outResponse, MemorySegment.NULL)
      val replacement =
        callback.transform(
          ResourceTransformRequest(ResourceKind.fromNative(rawKind), copyCString(url))
        )
      if (!replacement.isNullOrEmpty()) {
        if ('\u0000' in replacement) return MaplibreStatus.INVALID_ARGUMENT.nativeCode
        return NativeAccess.setResourceTransformResponseUrl(outResponse, replacement)
      }
      MaplibreStatus.OK.nativeCode
    } catch (_: IllegalArgumentException) {
      MaplibreStatus.INVALID_ARGUMENT.nativeCode
    } catch (_: Throwable) {
      MaplibreStatus.NATIVE_ERROR.nativeCode
    } finally {
      lease.close()
    }
  }

  fun checkCanClose() = gate.checkCanClose()

  fun isClosedForTesting(): Boolean = gate.isClosedForTesting()

  override fun close() = gate.close()

  @Suppress("UNUSED_PARAMETER")
  fun release(userData: MemorySegment) {
    HandleLeakCleaner.releaseNativeCallbackRoot(this)
    close()
  }

  private fun copyCString(address: MemorySegment): String {
    if (address == MemorySegment.NULL) {
      return ""
    }
    var length = 0L
    while (address.reinterpret(length + 1).get(ValueLayout.JAVA_BYTE, length) != 0.toByte()) {
      length++
    }
    return String(address.reinterpret(length).toArray(ValueLayout.JAVA_BYTE), Charsets.UTF_8)
  }

  private companion object {
    private val callbackDescriptor = mln_resource_transform_callback.descriptor()
  }
}
